### PR TITLE
add support for more numeric types

### PR DESCRIPTION
**Pull request by sritchie**
This PR is focused on adding new numeric implementations in Clojurescript, and adding more tests... but of course I let myself get a little wild, here.

- converts `sicmutils.modint` and tests to `cljc`
- converts `ModInt` to generic operations, so that we can create instances from other integral types 
- abstracts the integral and floating point tests out to functions that we can call for all the numeric types
- adds `Value` implementations and generic operation support for:
  - `js/BigInt`
  - `goog.math.Long`, the Google Closure library's 64 bit integer
  - `goog.math.Integer`, the arbitrary-precision datatype in the Closure library.
- adds a `sicmutils.generic/modulo` method that builds on top of `g/remainder`

Next up, my plan is to:

- try out this change to make exact Clojurescript numbers return as `::integral`, to see if that helps us get a little more clarity in the type system
- fill in more of the numeric tower. My suggestion is that we promote

`goog.math.Long` -> `goog.math.Integer`, and
`goog.math.Integer` -> `js/BigInt`...

under the assumption that native types are better to keep around... but maybe this is wrong. What do you think? Maybe it's best to assume that someone has a good reason for using `goog.math.Integer` if they added it and keep that.

Or maybe, now that it's easier to add these, I should move on!

### CODE REVIEW

**Comment by sritchie** [comment]
@littleredcomputer, this is ready for review!

The Travis CI build is redundant, so you could turn that off if you want. It's failing now because it's running an old version of Node that doesn't yet have `js/BigInt` support. That showed up in 2018, so I think it's fine to rely on this API.

File: src/sicmutils/numbers.cljc [hunk 1]
```diff
@@ -29,23 +29,37 @@
...
+            #?(:cljs goog.math.Integer)
+            #?(:cljs goog.math.Long))
   #?(:clj
      (:import (clojure.lang BigInt Ratio))))
 
-#?(:cljs (def ^:private ratio? (constantly false)))
-
+;; Smaller inheritance tree to enabled shared implementations between numeric
+;; types that represent mathematical integers.
+(derive u/inttype ::integral)
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply src/sicmutils/numbers.cljc:40]: I do agree. This comes up in cases like `expt` where the simplifier will see this as iterated multiplication with an integral argument and something completely opaque in every other case. That's on the input side. Simplifications involving `expt` also need the integral check.

File: src/sicmutils/numbers.cljc [hunk 2]
```diff
@@ -29,23 +29,37 @@
...
+(derive u/longtype ::integral)
 
+#?(:cljs
+   (derive js/BigInt ::integral)
 
-(defn ^:private define-unary-operation
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply src/sicmutils/numbers.cljc:40]: 👍

File: src/sicmutils/value.cljc [hunk 3]
```diff
@@ -90,6 +93,67 @@
...
+         (.compare this other)))))
+
+#?(:cljs
+   ;; Clojurescript-specific implementations of Value.
+   (let [big-zero (js/BigInt 0)
+         big-one (js/BigInt 1)]
+
+     (extend-protocol Value
+       js/BigInt
+       (nullity? [x] (js*  "~{} == ~{}" big-zero x))
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply src/sicmutils/value.cljc:128]: Fascinating. I really have to dig into this.

**Pull request merged**